dag_in_context/src/pretty_print.rs

```diff
@@ -0,0 +1,831 @@
+                    self.assign_fresh_var(lhs);
+                    self.assign_fresh_var(rhs);
+                }
+                Expr::If(cond, input, then, els) => {
+                    self.assign_fresh_var(cond);
+                    self.assign_fresh_var(input);
+                    self.assign_fresh_var(then);
+                    self.assign_fresh_var(els);
+                    try_insert_fresh(expr_symbol, "if".into(), self);
+                }
+                Expr::Switch(cond, input, branch) => {
+                    self.assign_fresh_var(cond);
+                    self.assign_fresh_var(input);
+                    branch.iter().for_each(|expr| self.assign_fresh_var(expr));
+                    try_insert_fresh(expr_symbol, "switch".into(), self);
+                }
+                Expr::DoWhile(input, body) => {
+                    self.assign_fresh_var(input);
+                    self.assign_fresh_var(body);
+                    try_insert_fresh(expr_symbol, "dowhile".into(), self);
+                }
+                Expr::Arg(ty, assum) => {
+                    try_insert_fresh(Symbols::Type(ty.clone()), ty.abbrev(), self);
+                    handle_assum(assum, self);
+                    try_insert_fresh(
+                        Symbols::Assumption(Assumption::to_ref(assum)),
+                        assum.abbrev(),
+                        self,
+                    );
+                }
+                Expr::Function(_, tyin, tyout, body) => {
+                    try_insert_fresh(Symbols::Type(tyin.clone()), tyin.abbrev(), self);
+                    try_insert_fresh(Symbols::Type(tyout.clone()), tyout.abbrev(), self);
+                    self.assign_fresh_var(body);
+                }
+                Expr::Symbolic(_) => panic!("no symbolic should occur when assigning freshvar"),
+            }
+        }
+    }
+
+    fn fold_expr(
+        &mut self,
+        expr: &RcExpr,
+        fold_when: &dyn Fn(usize, usize) -> bool,
+        to_rust: bool,
+    ) -> Expr {
+        let old_expr_addr = Rc::as_ptr(expr);
+        let fold = |pp: &mut PrettyPrinter, new_expr: schema::Expr| {
+            let fresh_var = pp.symbols.get(&Symbols::Expr(old_expr_addr)).unwrap();
+            if !pp.table.contains_key(fresh_var) {
+                pp.log.push(fresh_var.into());
+                pp.table.insert(fresh_var.into(), SymbolLog::Expr(new_expr));
+            }
+            Expr::Symbolic(fresh_var.into())
+        };
+
+        let num_shared = Rc::strong_count(expr);
+        let fold_or_plain = |pp: &mut PrettyPrinter, new_expr: Expr| {
+            let size = &new_expr
+                .to_string()
+                .replace(&['(', ')', ' ', ','][..], "") //don't count those char when computing size
+                .len();
+            if fold_when(num_shared, *size) {
+                fold(pp, new_expr)
+            } else {
+                new_expr
+            }
+        };
+
+        fn handle_assum(
```

Comment:
Your method names need to be more descriptive
Perhaps rename to "refactor_shared_assum"